/*
 * MIT License
 *
 * Copyright (c) 2026 Roman Bielyi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

syntax = "proto3";

package api.v1;

option go_package = "api/v1/gen/;apiv1";

import "api/v1/common.proto";
import "google/protobuf/timestamp.proto";

enum EphemType {
  EphemPlaceholder = 0;
  EphemEci = 1;
  EphemJ2K = 2;
}

message EphemTimeGrid {
  double time_start_ds50 = 1;
  google.protobuf.Timestamp time_start_utc = 2;
  double time_end_ds50 = 3;
  google.protobuf.Timestamp time_end_utc = 4;
  oneof time_step_type {
    double known_time_step_ds50 = 5;
    string known_time_step_period = 6;
    bool dynamic_time_step = 7;
  }
}

message EphemTask {
  int64 task_id = 1;
  Satellite sat = 2;
  EphemTimeGrid time_grid = 3;
}

message EphemRequest {
  int64 req_id = 1;
  EphemType ephem_type = 2;
  EphemTimeGrid common_time_grid = 3;
  repeated EphemTask tasks = 4;
}

message EphemOut {
  int64 task_id = 1;
  repeated EphemerisData ephem_data = 2;
  int64 ephem_points_count = 3;
}

message EphemResponse {
  int64 req_id = 1;
  int64 stream_id = 2;
  int64 stream_chunk_id = 3;
  EphemOut result = 4;
}